library/network/src/lib/network/firewalld.rb

```diff
@@ -0,0 +1,359 @@
+# encoding: utf-8
+#
+# ***************************************************************************
+#
+# Copyright (c) 2016 SUSE LLC.
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of version 2 or 3 of the GNU General
+# Public License as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE about this file by physical or electronic mail,
+# you may find current contact information at www.suse.com
+#
+# ***************************************************************************
+#
+# File: lib/network/firewalld.rb
+# Summary:  FirewallD configuration API
+# Authors:  Karol Mroz <kmroz@suse.de>, Markos Chandras <mchandras@suse.de>
+#
+
+require "yast"
+
+module Firewalld
+  # exception when firewall-cmd fails
+  class FirewallCMDError < StandardError
+  end
+
+  # Execute firewalld commands using firewall-cmd
+  class FWCmd
+    include Yast::Logger
+
+    BASH_SCR_PATH = Yast::Path.new(".target.bash_output")
+    # Base firewall-cmd command
+    COMMAND = "LANG=C firewall-cmd"
+
+    attr_reader :option_str
+
+    def initialize(option_str)
+      @option_str = option_str
+    end
+
+    def command
+      "#{COMMAND} #{option_str}".strip.squeeze(" ")
+    end
+
+    # @param output [Boolean] Whether command output is desirable or not
+    # @return [String] Firewalld command output
+    # @return [Boolean] FIrewalld command exit code
+    def fwd_result(output = true)
+      cmd_result = Yast::SCR.Execute(BASH_SCR_PATH, command)
+
+      # See firewall-cmd manpage for exit codes. Not all of them justify an
+      # exception. 0 and 1 can be used as true and false respectively.
+      case cmd_result["exit"]
+      when 0, 1
+        if output
+          cmd_result["stdout"]
+        else
+          log.debug "#{command} returned: #{cmd_result["stdout"]}"
+          cmd_result["exit"].zero? ? true : false
+        end
+      else
+        raise FirewallCMDError, "Calling firewall-cmd (cmd: #{command}) failed: #{cmd_result["stderr"]}"
+      end
+    end
+  end
```

Comment:
when I see this code I suggest to look at https://github.com/yast/yast-yast2/blob/9c7caad3f3332f607937e9b236d90c202e4b1de9/library/system/src/lib/yast2/execute.rb and cheetah API that is used, which will provide easier API for you including good exception and nice logging - https://github.com/openSUSE/cheetah